baybe/objectives/desirability.py

```diff
@@ -0,0 +1,135 @@
+"""Functionality for desirability objectives."""
+
+from collections.abc import Sequence
+from functools import partial
+from typing import Callable, Union
+
+import cattrs
+import numpy as np
+import numpy.typing as npt
+import pandas as pd
+from attrs import define, field
+from attrs.validators import deep_iterable, instance_of, min_len
+from typing_extensions import TypeGuard
+
+from baybe.objectives.base import Objective
+from baybe.objectives.enum import Scalarization
+from baybe.targets.base import Target
+from baybe.targets.numerical import NumericalTarget
+from baybe.utils.basic import to_tuple
+from baybe.utils.numerical import geom_mean
+
+
+def _normalize_weights(weights: Sequence[Union[float, int]]) -> tuple[float, ...]:
+    """Normalize a collection of weights such that they sum to 1.
+
+    Args:
+        weights: The un-normalized weights.
+
+    Raises:
+        ValueError: If any of the weights is non-positive.
+
+    Returns:
+        The normalized weights.
+    """
+    array = np.asarray(cattrs.structure(weights, tuple[float, ...]))
+    if not np.all(array > 0.0):
+        raise ValueError("All weights must be strictly positive.")
+    return tuple(array / array.sum())
+
+
+def _is_all_numerical_targets(
+    x: tuple[Target, ...], /
+) -> TypeGuard[tuple[NumericalTarget, ...]]:
```

Comment:
Just typing-related things. Instead of me explaining, I think it's better if you just look here: https://mypy.readthedocs.io/en/stable/type_narrowing.html

Please resolve if clear

Reply:
assuming this is now clear, I'll resolve